core/src/main/java/com/linecorp/armeria/client/AbstractClientOptionsBuilder.java

```diff
@@ -368,6 +373,66 @@ public AbstractClientOptionsBuilder followRedirects(RedirectConfig redirectConfi
         return option(REDIRECT_CONFIG, requireNonNull(redirectConfig, "redirectConfig"));
     }
 
+    /**
+     * Adds the specified {@link ClientRequestContext} customizer function so that
+     * it customizes the context in the thread that initiated the client call.
+     * The given customizer function is evaluated before the customizer function specified by
+     * {@link Clients#withContextCustomizer(Consumer)}.
+     *
+     * <pre>{@code
+     * static final ThreadLocal<String> USER_ID = new ThreadLocal<>();
+     * static final AttributeKey<String> USER_ID_ATTR = AttributeKey.valueOf("USER_ID");
+     * ...
+     * MyClientStub client =
+     *     Clients.builder(...)
+     *            .contextCustomizer(ctx -> {
+     *                // This customizer will be invoked from the thread that initiates a client call.
+     *                ctx.setAttr(USER_ID_ATTR, USER_ID.get());
+     *            })
+     *            .build(MyClientStub.class);
+     *
+     * // Good:
+     * // The context data is set by the thread that initiates the client call.
+     * USER_ID.set("user1");
+     * client.executeSomething1(..);
+     * ...
+     * client.executeSomethingN(..);
+     * // Should clean up the thread local storage.
+     * USER_ID.remove();
```

Comment:
Could use `try-finally`:
```suggestion
     * try {
     *     client.executeSomething1(..);
     *     ...
     *     client.executeSomethingN(..);
     * } finally {
     *     // Should clean up the thread local storage.
     *     USER_ID.remove();
     * }
```